packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1922 @@
+  return { subjects: { anonymous: true, persist: true } } as StateMetadata<
+    PermissionControllerState<Permission>
+  >;
+}
+
+/**
+ * Get the default state of the {@link PermissionController}.
+ *
+ * @template Permission - The controller's permission type union.
+ */
+function getDefaultState<Permission extends PermissionConstraint>() {
+  return { subjects: {} } as PermissionControllerState<Permission>;
+}
+
+/**
+ * Gets the state of the {@link PermissionController}.
+ */
+export type GetPermissionControllerState = {
+  type: `${typeof controllerName}:getState`;
+  handler: () => PermissionControllerState<PermissionConstraint>;
+};
+
+/**
+ * Gets the names of all subjects from the {@link PermissionController}.
+ */
+export type GetSubjects = {
+  type: `${typeof controllerName}:getSubjectNames`;
+  handler: () => (keyof PermissionControllerSubjects<PermissionConstraint>)[];
+};
+
+/**
+ * Checks whether the specified subject has any permissions.
+ */
+export type HasPermissions = {
+  type: `${typeof controllerName}:hasPermissions`;
+  handler: GenericPermissionController['hasPermissions'];
+};
+
+/**
+ * Clears all permissions from the {@link PermissionController}.
+ */
+export type ClearPermissions = {
+  type: `${typeof controllerName}:clearPermissions`;
+  handler: () => void;
+};
+
+// TODO: Implement all desired actions
+/**
+ * The {@link ControllerMessenger} actions of the {@link PermissionController}.
+ */
+export type PermissionControllerActions =
+  | ClearPermissions
+  | GetPermissionControllerState
+  | GetSubjects
+  | HasPermissions;
+
+/**
+ * The generic state change event of the {@link PermissionController}.
+ */
+export type PermissionControllerStateChange = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [PermissionControllerState<PermissionConstraint>, Patch[]];
+};
+
+/**
+ * The {@link ControllerMessenger} events of the {@link PermissionController}.
+ *
+ * The permission controller only emits its generic state change events.
+ * Consumers should use selector subscriptions to subscribe to relevant
+ * substate.
+ */
+export type PermissionControllerEvents = PermissionControllerStateChange;
+
+/**
+ * The external {@link ControllerMessenger} actions available to the
+ * {@link PermissionController}.
+ */
+type AllowedActions =
+  | AddApprovalRequest
+  | HasApprovalRequest
+  | AcceptApprovalRequest
+  | RejectApprovalRequest;
+
+/**
+ * The messenger of the {@link PermissionController}.
+ */
+export type PermissionControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  PermissionControllerActions | AllowedActions,
+  PermissionControllerEvents,
+  AllowedActions['type'],
+  never
+>;
+
+/**
+ * A generic {@link PermissionController}.
+ */
+export type GenericPermissionController = PermissionController<
+  PermissionSpecificationConstraint,
+  CaveatSpecificationConstraint
+>;
+
+/**
+ * Options for {@link PermissionController.grantPermissions}.
```

Comment:
I think it would be easier to read if this was inlined. Having to scroll back and forth between the function and this type was inconvenient - I ended up creating a separate editor pane just to keep them both visible at once.

Reply:
11e318e